examples/4_advanced_1/main.go

```diff
@@ -31,13 +32,34 @@ func main() {
 
 	gr := gornir.New().WithInventory(inv).WithLogger(log).WithRunner(rnr)
 
-	results := make(chan *gornir.JobResult, len(gr.Inventory.Hosts))
+	// Open an SSH connection towards the devices
+	results, err := gr.RunSync(
+		&connection.SSHOpen{},
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+	output.RenderResults(os.Stdout, results, "Connecting to devices via ssh", true)
+
+	// defer closing the SSH connection we just opened
+	defer func() {
+		results, err = gr.RunSync(
+			&connection.SSHClose{},
+		)
+		if err != nil {
+			log.Fatal(err)
+		}
```

Comment:
The error on an individual host is propagated through the result object. The only error a runner should return should be related to the runner itself. For instance, I have my own custom runners that check for conditions in the network before doing anything, if the conditions aren't met the runner will return an error saying it can't perform the action (i.e., you have a bunch of spines and you are going to be executing a given task one by one to ensure there is no downtime and prior to each execution you consult prometheus to verify at least 50% of the spines are healthy).

Does this make sense?

Reply:
Oh, and the reason I prefer to encode that logic in the runner instead of via some grouped task is because the same workflows could be executed this way using different strategies depending on the conditions. For instance, on a new site I might run workflow A in parallel because the site is not live yet but in a live site I might run that same workflow using a more caring strategy. In the end, the runner should dictate how things are done and gated and the task itself should only care about the actions to perform on a given device.